chainermn/communicators/pure_nccl_communicator.py

```diff
@@ -62,76 +65,113 @@ def _init_comms(self):
     def broadcast_data(self, model):
         _communication_utility.broadcast_naive(self.mpi_comm, model)
 
-    def allreduce_grad(self, model, stream=None):
-        self._init_comms()
-        if stream is None:
-            stream = chainer.cuda.Stream.null
+    def allreduce_grad(self, model):
+        stream = chainer.cuda.Stream.null
+        self.allreduce_grad_async(model, stream)
 
+    def allreduce_grad_async(self, model, stream):
+        self._init_comms()
         params = _memory_utility.extract_params(model)
         grad_dtype = _get_param_grad_dtype(params[0])
         if self.allreduce_grad_dtype is None:
             allreduce_grad_dtype = grad_dtype
         else:
             allreduce_grad_dtype = self.allreduce_grad_dtype
         n_elems = sum(param.grad.size for param in params)
-        self._assign(grad_dtype, allreduce_grad_dtype, n_elems)
-        self._pack_params_to_buffer(params, grad_dtype, allreduce_grad_dtype,
-                                    n_elems)
-        if stream != chainer.cuda.Stream.null:
+        needs_sync = self._assign(grad_dtype, allreduce_grad_dtype, n_elems)
+        if stream != chainer.cuda.Stream.null and needs_sync:
             chainer.cuda.Stream.null.synchronize()
+
+        self._pack_params_to_buffer(params, grad_dtype, allreduce_grad_dtype,
+                                    n_elems, stream)
         self.nccl_comm.allReduce(self.gpu_allreduce_buffer_a.ptr(),
                                  self.gpu_allreduce_buffer_b.ptr(), n_elems,
                                  _get_nccl_type_id(allreduce_grad_dtype),
-                                 nccl.NCCL_SUM, stream.ptr)
-        if stream != chainer.cuda.Stream.null:
-            stream.synchronize()
```

Comment:
Ditto why this can be removed.

Reply:
Because, I added synchronize() at the end of allreduce_grad in DoubleBufferingOptimizer.
https://github.com/chainer/chainermn/pull/216/files#diff-5ac36de863ea63673cdc464693ae1accR131